_posts/2021-01-23-modeling-ui-state.md

```diff
@@ -0,0 +1,196 @@
+
+- an enum - `enum class Color { RED, YELLOW, GREEN }` - this has a cardinality = 3 which is exactly the same as the possible valid values of the problem. Illegal values are now impossible, so there is no need for tests that check data validity.
+
+By modelling the data in a way that rules out illegal values, the resulting code will end up shorter, more clear and easier to test. 
+
+> Make sure the set of possible values in code match the set of valid values of the problem and a lot of problems disappear.
+
+## Exposing State from a ViewModel
+
+The problem: build an app that calls a traffic light endpoint and shows the color of the traffic light (red, yellow or green). During the network show a `ProgressBar`, on success show a `View` with the color and in case of an error show a `TextView` with generic text. There is no retry, exactly one view is show at a time.
+
+```kotlin
+class TrafficLightViewModel : ViewModel() {
+
+    val state: LiveData<TrafficLightState> = TODO()
+}
+```
+
+I will represent the color as the enum with three values. How should be the type of `TrafficLightState` look like?
+
+```kotlin
+data class TrafficLightState(
+    val isLoading: Boolean, 
+    val isError: Boolean, 
+    val color: Color?
+)
+```
+
+one way is a data class with three properties. However this has possible invalid states e.g. `TrafficLightState(true, true, Color.RED)`. Both error and loading are true (show the `ProgressBar` or the error `TextView` ?) but we also have a Color? The cardinality of `TrafficLightState` is 2 (Boolean) * 2 (Boolean) * 4 (Color?) = 16.
+
+Another approach is using a sealed class:
+
+```kotlin
+sealed class TrafficLightState
+object Loading : TrafficLightState()
+object Error : TrafficLightState()
+data class Success(val color: Color) : TrafficLightState()
+```
+
+which has a cardinality of 1 (Loading) + 1 (Error) + 3 (Success) = 5 which exactly matches the possible states of the problem:
+
+- during the network call: `Loading` and show the `ProgressBar`
+- if the network call fails: `Error` and show a `TextView`
+- on success: `Success` and show a view with the corresponding color
+
+What about using multiple observable properties:
+
+```kotlin
+class TrafficLightViewModel : ViewModel() {
+
+    val loading: LiveData<Boolean> = TODO()
+
+    val error: LiveData<Boolean> = TODO()
+
+    val color: LiveData<Color> = TODO()
+}
+```
+
+this has the same cardinality as the data class `TrafficLightState` approach - 16 and makes invalid values possible.
+
+These are all valid approaches and when designing your model but only one of them matches exactly the possible valid states of the problem. To eliminate bugs and reduce the number of tests, use that one.
```

Comment:
Recommended something like: "These are all valid approaches but, when designing your model and only one of the possibilities matches the real values of the actual problem, in order to eliminate bugs and reduce the number of tests, use that one."